### PR TITLE
add mxnet

### LINKED ISSUE #4447: Package mxnet
@pjmdtw would you be intersted in contributing your mxnet package to conda-forge ?
https://github.com/pjmtdw/mxnet-conda-builder

conda-forge already has xgboost and it would be excellent to have mxnet too.

### PR DESCRIPTION
**Pull request by ngam**
Fixes #4447 
Closes #14832 

---

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by ngam** [comment]
@barry-jin, this is more or less ready (pending CI). Could you have a look and see if there are any issues? 

For the standard linux and darwin, I just copied them straight from config/. For the cuda ones, I copied them from config/distribution. We will likely have to do more optimization later on; and perhaps set this up in a proper matrix instead of this. But that can wait until we get this past here. Let me know what you think.

**Comment by ngam** [comment]
cc @jakirkham if you want to have a look here.

**Comment by ngam** [comment]
@jakirkham 

Any help here? Thanks!

```
-- Looking for a CUDA compiler - NOTFOUND
CMake Warning at CMakeLists.txt:121 (message):
  CMAKE_CUDA_COMPILER guessed: /usr/local/cuda/bin/nvcc

  Please fix your cuda installation:
  https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#mandatory-post


-- The CUDA compiler identification is unknown
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - failed
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc - broken
CMake Error at /home/conda/staged-recipes/build_artifacts/mxnet_1642711491483/_build_env/share/cmake-3.21/Modules/CMakeTestCUDACompiler.cmake:56 (message):
  The CUDA compiler

    "/usr/local/cuda/bin/nvcc"

  is not able to compile a simple test program.

  It fails with the following output:

```

**Comment by ngam** [comment]
Current problems / need-help areas:

- [ ] tighter pinning of cudnn (getting an error about `cudnn undefined`)? --- cc @jakirkham 
- [ ] how to actually do the python call at the very end of build.sh? Why isn't it not working as expected? @conda-forge/help-python 
- [ ] why is the splitting not working well?
- [ ] how best to handle the different layers of BLAS, OpenMP, etc. (ideally, I would want to have them as variants, but I am not sure how variants would work in this case; for now, I am going for the standard configs upstream, but we definitely could do better) cc @isuruf

**Comment by ngam** [comment]
@conda-forge/staged-recipes ready for review

File: recipes/mxnet/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    - libgomp  # [linux]
+    - llvm-openmp  # [osx]
+    - openblas
+    - mkl  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - mkl-include  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - opencv
+    - onednn
+    - cudnn  # [cuda_compiler_version != "None"]
+    - nccl  # [cuda_compiler_version != "None"]
+    - numpy >=1.17
```
Comment (hmaarrfk) [inline recipes/mxnet/meta.yaml:89]: Numpy is quite forward compatible, but it isn't really backward compatible. Doing this kind of pinning uses 
```
2022-01-22T08:20:42.7430135Z   + numpy                            1.21.5  py37hf2998dd_0        conda-forge/linux-64       6 MB
```
which makes it only compatible with numpy 1.21.5

Please follow
https://conda-forge.org/docs/maintainer/knowledge_base.html#building-against-numpy

File: recipes/mxnet/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    - {{ pin_compatible('libgomp') }}  # [linux]
+    - {{ pin_compatible('llvm-openmp') }}  # [osx]
+    - openblas
+    - {{ pin_compatible('mkl') }}  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - {{ pin_compatible('mkl-include') }}  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - {{ pin_compatible('opencv') }}
+    - {{ pin_compatible('onednn') }}
+    - {{ pin_compatible('cudnn') }}  # [cuda_compiler_version != "None"]
+    - {{ pin_compatible('nccl') }}  # [cuda_compiler_version != "None"]
+    - numpy >=1.17
```
Comment (hmaarrfk) [inline recipes/mxnet/meta.yaml:107]:
```suggestion
    - {{ pin_compatible('numpy') }}
```

No matter what your pinning is above, this is the correct line to include.

File: recipes/mxnet/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    - requests >=2.20.0,<3
+    - python-graphviz >=0.8.1,<0.9.0
+    - python_version
+    - {{ pin_compatible('libjpeg-turbo') }}
```
Comment (hmaarrfk) [inline recipes/mxnet/meta.yaml:111]: This is going to cause inter-operability problems. You should simply build with jpeg. Do you need turbo? or is it just for performance?

File: recipes/mxnet/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    - libjpeg-turbo
+    - python
+
+  run:
+    # ?
+    - {{ pin_compatible('libgomp') }}  # [linux]
+    - {{ pin_compatible('llvm-openmp') }}  # [osx]
+    - openblas
+    - {{ pin_compatible('mkl') }}  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
+    - {{ pin_compatible('mkl-include') }}  # [(target_platform != "osx-arm64") and (cuda_compiler_version == "None")]
```
Comment (hmaarrfk) [inline recipes/mxnet/meta.yaml:102]: Do you need development headers?

File: recipes/mxnet/meta.yaml (hunk 1 above)
Comment (ngam) [reply recipes/mxnet/meta.yaml:89]: So, should I just `numpy ==1.17.*` or should I just drop the pinning here?

File: recipes/mxnet/meta.yaml (hunk 3 above)
Comment (ngam) [reply recipes/mxnet/meta.yaml:111]: I believe they explicitly require turbo, but I can look deeper into it. It is actually _built_ against libjpeg-turbo and I am not really sure if it is a runtime dep. I added all this pinning because when I was testing locally, I would run into problems like "can't find library libopencv.xxx.405" so I thought ooops, I gotta pin them all to start and then relax them later. What's your recommendation here? We potentially will have all sorts of problematic pinning because these are used in the building, not merely as a flexible runtime deps and so this is quite challenging for me to understand as a beginner!

File: recipes/mxnet/meta.yaml (hunk 4 above)
Comment (ngam) [reply recipes/mxnet/meta.yaml:102]: There is an explicit line in build.sh for the MKL include, but I haven't really checked if this passes without this and I am the whole onednn--mkl building is very confusing :( 

I think mkl-include can be dropped, but I will want to check that first

**Comment by ngam** [comment]
@hmaarrfk, this is going to be a challenging and interesting feedstock to maintain --- while I am only running some of the "supported" distributions here (config files), the whole thing upstream is really flexible and we can potentially do a lot of interesting mixes for our purposes and help them out with distributing and optimizing their build. For example, currently, I believe at least some of onednn gets built --- even though it looks for it at the start to ensure compatibility, which is very confusing!

Anyway, would you like me to add you as a maintainer here? This way you'd be on the tensorflow, pytorch, and mxnet --- the whole axis lol. And you will teach me all the trick :) Copying @h-vetinari to gauge interest in joining the maintainership too

File: recipes/mxnet/meta.yaml (hunk 1 above)
Comment (ngam) [reply recipes/mxnet/meta.yaml:89]: I actually don't if numpy needs to be in this section... I will need to check that too!

Comment (hmaarrfk) [reply recipes/mxnet/meta.yaml:89]: just drop the pin. it will get a newer number from the config.

File: recipes/mxnet/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    portable and lightweight, scaling effectively to multiple GPUs and multiple
+    machines. MXNet is also more than a deep learning project. It is also a
+    collection of blue prints and guidelines for building deep learning systems,
+    and interesting insights of DL systems for hackers.
+  doc_url: https://mxnet.incubator.apache.org/
+  dev_url: https://github.com/apache/incubator-mxnet
+
+extra:
+  recipe-maintainers:
+    - barry-jin
```
Comment (hmaarrfk) [inline recipes/mxnet/meta.yaml:146]: This person should chime in before we can merge. I suggest you add them after in the feedstock itself. instead of now at the beginning.

Comment (ngam) [reply recipes/mxnet/meta.yaml:146]: @barry-jin indicated willingness (also a maintainer upstream). See above for the "like" and here https://github.com/conda-forge/staged-recipes/issues/4447#issuecomment-1015956430 but happy to remove for now and tidy up things later. 

Btw, I couldn't figure out a clean way to get the library separated. The python package itself can be noarch and can be built separately. This is one main thing I wanna do once we get this into a feedstock. Another thing is being involved with niche (?) dependencies so that we can avoid "vendoring" them (to be clear, we are not vendoring much, but still I would like to figure out a way to get dlpack, dmlc, nnvm (libtvm) and mshadow from elsewhere.

File: recipes/mxnet/meta.yaml (hunk 1 above)
Comment (ngam) [reply recipes/mxnet/meta.yaml:89]: OK, will remove. I now think numpy should only be in run... testing this takes forever, that's why I am hesitating to keep going back and forth, but I will try tweaking things around and tidying up locally first.

Comment (ngam) [reply recipes/mxnet/meta.yaml:89]: Thanks a lot! 🥇

**Comment by ngam** [comment]
deps need fixing

```
   INFO: sysroot: '/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/' files: '['usr/share/man/mann/zero.n', 'usr/share/man/mann/yencode.n', 'usr/share/man/mann/yaml.n', 'usr/share/man/mann/xsxp.n']'
   INFO (mxnet,mxnet/libmxnet.dylib): Needed DSO lib/libmkl_intel_ilp64.1.dylib found in conda-forge::mkl-2021.4.0-h89fa619_689
   INFO (mxnet,mxnet/libmxnet.dylib): Needed DSO /Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/lib/libSystem.B.dylib found in $SYSROOT
   INFO (mxnet,mxnet/libmxnet.dylib): Needed DSO lib/libturbojpeg.0.dylib found in conda-forge::libjpeg-turbo-2.1.1-h0d85af4_0
  ERROR (mxnet,mxnet/libmxnet.dylib): Needed DSO lib/libopencv_imgcodecs.405.dylib found in ['libopencv']
  ERROR (mxnet,mxnet/libmxnet.dylib): .. but ['libopencv'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
  ERROR (mxnet,mxnet/libmxnet.dylib): Needed DSO lib/libopencv_imgproc.405.dylib found in ['libopencv']
  ERROR (mxnet,mxnet/libmxnet.dylib): .. but ['libopencv'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
  ERROR (mxnet,mxnet/libmxnet.dylib): Needed DSO lib/libopencv_core.405.dylib found in ['libopencv']
  ERROR (mxnet,mxnet/libmxnet.dylib): .. but ['libopencv'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
  ERROR (mxnet,mxnet/libmxnet.dylib): Needed DSO lib/libprotobuf.30.dylib found in ['libprotobuf']
  ERROR (mxnet,mxnet/libmxnet.dylib): .. but ['libprotobuf'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
   INFO (mxnet,mxnet/libmxnet.dylib): Needed DSO lib/libc++.1.dylib found in conda-forge::libcxx-12.0.1-habf9029_1
WARNING (mxnet): interpreter (Python) package conda-forge::python-3.8.12-h17280f6_2_cpython in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): interpreted library (Python) package conda-forge::python_version-0.0.2-pyhd8ed1ab_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): interpreted library (Python) package conda-forge::python-graphviz-0.8.4-py_1003 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): run-exports library package conda-forge::llvm-openmp-11.1.0-hda6cdc1_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): interpreted library (Python) package conda-forge::pip-21.3.1-pyhd8ed1ab_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): run-exports library package conda-forge::libgfortran5-9.3.0-h6c81a4c_23 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): run-exports library package conda-forge::onednn-2.5.2-omp_h4900ede_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): plugin library (Python) package conda-forge::numpy-1.22.0-py38h9d72dae_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (mxnet): interpreted library (Python) package conda-forge::requests-2.27.1-pyhd8ed1ab_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)

```

**Comment by ngam** [comment]
While this is actually more or less ready, I'd rather spend a little more time trying to split it into a library and python pkg instead of this. Converting back to draft, will ping again soon.

**Comment by h-vetinari** [comment]
@ngam, I missed this PR at the time, and I don't see much context for its closure (except that I remember you don't like having open PRs...?). What would you think about giving this one another shot?
